### ISSUE TITLE
jQuery DataTables and Editor - the evil

### ISSUE THREAD

**Issue by uniquejava** [uniquejava/blog#124]
https://datatables.net/ and https://editor.datatables.net/

## 比较一下jsgrid和datatables的Getting Started页面

[DataTables Getting Started](https://editor.datatables.net/manual/getting-started)

[jsgrid Getting Started](http://js-grid.com/getting-started/)

## 吐槽
DataTables所有的列都要指定fields和columns (N多重复代码)
data, visible, className要定义在column里边, name, type要定义在fields里边, WTF?

jsgrid只要定义fields:  name, visible和type等一个field轻松搞定, no brain-fuck.

DataTables的下拉框真的难用, 文档在tech notes里边(难找吧!)
https://datatables.net/manual/tech-notes/11
下拉字段是下面这样的形式
fields.push({name: 'FieldName', type: 'select', options:  [{label:'', value:''}, ...] });
其中options必须是这种指定格式. 需要提前转换, jsgrid没有这个限制(无需转换)..

DataTables的editor还是收费的. 依赖一大堆组合, buttons是一个js, select又是一个js, 到底要下载哪些js分分钟把你搞晕

DataTables出了错竟然是弹出alert(这也太不友好了, 在console.log中打印出来不更好)
关键是warning类型的没有必要弹alert.

如何禁掉这个alert在此:
https://datatables.net/forums/discussion/28726/how-to-disable-the-warning-message

## references

https://datatables.net/forums/discussion/29339/change-ajax-url

https://editor.datatables.net/reference/option/ajax.data

https://editor.datatables.net/reference/field/select

**Comment by uniquejava** [comment]
## angular-datatables
angular-datatables官方文档: http://l-lin.github.io/angular-datatables/archives/

如果要使用bootstrap的样式, 见http://l-lin.github.io/angular-datatables/archives/#!/bootstrapIntegration这部分.

withDOM的文档: https://datatables.net/reference/option/dom

如果要使用server side分页, 而不能用angular way, 只能用ajax.

**Comment by uniquejava** [comment]
## Fixed table header(scroller)
1. `angular way`: 这种方法写html简单, 但是固定表头不能随窗口改变动态调整大小.
2. dt-columns, 代码长, 但是不会有resize的问题.

```js
    // Code for fixed table header
    vm.dtInstances = [];

    vm.dtOptions = DTOptionsBuilder.newOptions()
      .withScroller()
      .withOption('scrollY', 152); // 5 rows

    vm.dtInstCallback = function (dtInstance) {
      vm.dtInstances.push(dtInstance);

      // resize fixed table header on window resize
      var onResize = _.throttle(_.bind(dtInstance.rerender, dtInstance), 1000);

      function cleanUp() {
        angular.element($window).off('resize', onResize);
        console.log('cleanUp');
      }

      angular.element($window).on('resize', onResize);
      $scope.$on('$destroy', cleanUp);

    };

```

**Comment by uniquejava** [comment]
## server side pagination
我这写了一个工具类.
```js
(function () {
  'use strict';

  angular.module('myapp')
    .factory('utils', utils);

  /** @ngInject */
  function utils($http, $compile, $timeout, DTOptionsBuilder, config) {
    return {

      calcRowNumber: function (vm, meta) {
        return (vm.pageParams.page - 1) * vm.pageParams.limit + meta.row + 1;
      },

      useNgDirective: function (dtOptions, scope) {
        // see https://stackoverflow.com/questions/36911340/how-to-access-current-row-data-in-angular-datatables-using-templates
        dtOptions.withOption('createdRow', function (row, data, dataIndex) {
          // Create a new scope for each row, otherwise, data will
          // not be unique for each row because of data bindings
          var $scope = scope;

          var $newScope = $scope.$new(true);
          $newScope.row = data;

          // Pass any methods you are using in current scope
          $newScope.delete = $scope.delete;

          return $timeout(function () {
            // Recompiling so we can bind Angular directive to the DT
            return $scope.$apply($compile(angular.element(row).contents())($newScope));
          });
        });
      },

      enableServerSidePagination: function (vm, options) {
        options = options || {};
        options.limit = options.limit || 10;

        var dtOptions = DTOptionsBuilder.newOptions();
        dtOptions.withOption('ajax', function (data, callback, settings) {
          // make an ajax request using data.start and data.length
          var limit = data.length;
          var page = data.start / limit + 1;

          $http.get(config.api + options.url,
            {params: {page: page, limit: limit}})
            .success(function (res) {
              // map your server's response to the DataTables format and pass it to
              // DataTables' callback
              var pageParams = {
                total: res.total,
                page: page,
                limit: res.limit
              };

              vm.pageParams = pageParams;

              callback({
                recordsTotal: res.total,
                recordsFiltered: res.total,
                data: res.items
              });


            });
        });

        if(options.ng && options.scope) {
          this.useNgDirective(dtOptions, options.scope);
        }

        /*
          see https://datatables.net/reference/option/dom

          l - length changing input control
          f - filtering input
          t - The table!
          i - Table information summary
          p - pagination control
          r - processing display element

         */

        dtOptions
          .withOption('bPaginate', true)
          .withOption('bInfo', true)
          .withDOM('rtip')
          .withOption('serverSide', true)
          .withPaginationType('full_numbers')
          .withDataProp('data')
          .withDisplayLength(options.limit)
          .withBootstrap();

        vm.dtOptions = dtOptions;
      }

    };
  }

})();

```

用法如下:
```js
vm.dtInstance = {};

// 反注释这行使用服务端分页
//utils.enableServerSidePagination(vm, {url: '/categories/v1', ng: true, scope: $scope});

var dtOptions = DTOptionsBuilder.fromFnPromise(function () {
    return $http.get(config.api + '/categories/v1');
  }).withOption('sAjaxDataProp', 'data.items');

utils.useNgDirective(dtOptions, $scope);

dtOptions.withScroller().withOption('scrollY', 200); 
vm.dtOptions = dtOptions;

vm.dtColumns = [
  // DTColumnBuilder.newColumn(null).withTitle('#').renderWith(function (data, type, full, meta) {
  //   return utils.calcRowNumber(vm, meta);
  // }),

  DTColumnBuilder.newColumn('name').withTitle('Name').renderWith(function (data) {
    return '{{row.name|limitTo: 5}}';
  }),
  DTColumnBuilder.newColumn('createdDate')
    .withClass('text-center')
    .withTitle('Created Date')
    .renderWith(function (data, type, full, meta) {
      return $filter('date')(data, 'yyyy/MM/dd');
    }),
  DTColumnBuilder.newColumn('customProperty').withTitle('Custom Property').notVisible(),
  DTColumnBuilder.newColumn('id').withTitle('Link').renderWith(function () {
    return '<a ui-sref="index.some({id:row.id})">some link</a>';
  })
];
```

**Comment by uniquejava** [comment]
## datatables callback

在datatables渲染完table后会回调 fnDrawCallback, 可以在withOption中注册这个回调函数.
